clients/application-amqp/src/main/java/org/eclipse/hono/application/client/amqp/AmqpBasedCommandSender.java

```diff
@@ -0,0 +1,66 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2021 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.application.client.amqp;
+
+import java.util.Map;
+
+import org.eclipse.hono.application.client.CommandSender;
+import org.eclipse.hono.client.ServiceInvocationException;
+
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+import io.vertx.core.buffer.Buffer;
+
+/**
+ * An AMQP 1.0 based client that supports Hono's north bound operations to send commands.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/api/command-and-control/">
+ *      Command &amp; Control API for AMQP 1.0 Specification</a>
+ */
+public interface AmqpBasedCommandSender extends CommandSender {
+
+    /**
+     * Sends a <em>one-way command</em> to a device, i.e. there is no response from the device expected.
+     * <p>
+     * A device needs to be (successfully) registered before a client can upload
+     * any data for it. The device also needs to be connected to a protocol adapter
+     * and needs to have indicated its intent to receive commands.
+     *
+     * @param tenantId The tenant that the device belongs to.
+     * @param deviceId The device to send the command to.
+     * @param command The name of the command.
+     * @param contentType The type of the data submitted as part of the one-way command or {@code null} if unknown.
+     * @param data The input data to the command or {@code null} if the command has no input data.
+     * @param properties The headers to include in the one-way command message.
+     * @param sendMessageTimeout The maximum number of milliseconds to wait.
+     * @param context The currently active OpenTracing span context that is used to trace the execution of this
+     *            operation or {@code null} if no span is currently active.
+     * @return A future indicating the result of the operation:
+     *         <p>
+     *         If the one-way command was accepted, the future will succeed.
+     *         <p>
+     *         The future will fail with a {@link ServiceInvocationException} if the one-way command could 
+     *         not be forwarded to the device.
+     * @throws NullPointerException if any of tenantId, deviceId or command are {@code null}.
+     * @throws IllegalArgumentException if the sendMessageTimeout value is &lt; 0
+     */
+    Future<Void> sendOneWayCommand(
```

Comment:
is there anything AMQP specific about this method? If not, what do we need this interface for?

Reply:
This method has an extra parameter `sendMessageTimeout` and currently this method is used only in the IT test in the context of AMQP (explicitly `AmqpApplicationClient` is being used in the `IntegrationTestSupport` as below). Hence I added it to the AMQP specific interface to support that test.

```
amqpApplicationClient.sendOneWayCommand(tenantId, deviceId, command, contentType, payload, properties,
                requestTimeout, NoopSpan.INSTANCE.context())
```

Reply:
That I understood/saw. The question is: why does this need to be in an AMQP specific interface when there is nothing AMQP specific about it. Why can't we simply add it to `CommandSender`?

Reply:
Even though the `sendMessageTimeout` is not AMQP specific and since the above method is only being used in tests (involving AMQP), I didn't pull up to the `CommandSender`.

We can simply add it to the `CommandSender`. But I doubt that all the underlying implementations can support it. For example, in case of Kafka we won't be able to use the `sendMessageTimeOut` at each transaction level. Kafka offers timeout properties which can be set at the producer configuration. Kafka normally works by collecting messages and sending them in batches and there is no provision in vertx kafka client API to achieve this timeout for each message transaction. In that case, we can add a note to the _JavaDoc_ saying that this timeout may be used if the underlying implementation supports it. FMPOV adding such disclaimers reduces clarity in understanding how a API would behave with various implementations.

Reply:
> But I doubt that all the underlying implementations can support it.

FMPOV this is not about taking advantage of the underlying transport's capabilities but simply about providing the client the opportunity to limit the time to wait for the completion of the returned Future. Otherwise, the client would need to implement the timeout logic itself. It doesn't matter whether AMQP sends each message individually or Kafka sends it in batches. If the client is not willing to wait more than 1 second for the Future to complete, then it will probably resend the message once the second has passed. If that means that a client is never able to (successfully) send a message via Kafka, then either the Kafka producer config needs to be adapted or the client needs to increase the timeout. The same is true for AMQP, i.e. if the underlying messaging network takes more than a second to forward a message from the client to the device, then the client will run into the timeout.
Again, FMPOV there is nothing transport specific about this method. However, the Kafka transport can not determine if the command could be forwarded to the device because it writes the command to the message log instead of forwarding it to the protocol adapter.

Reply:
> FMPOV this is not about taking advantage of the underlying transport's capabilities but simply about providing the client the opportunity to limit the time to wait for the completion of the returned Future. Otherwise, the client would need to implement the timeout logic itself.

The client need not implement the timeout logic itself. The client already has this flexibility to limit the time to wait for completion but at the `ApplicationClient` level. In case of AMQP it is the `sendMessageTimeout` at the connection config. In case of Kafka it can be set using the producer config property `delivery.timeout.ms`. FMPOV if a user wants to change the existing value, then he/she can create a new client instance with the new timeout value.

> FMPOV there is nothing transport specific about this method

I agree that there is nothing transport specific. I can move it to the `CommandSender`. 

But in case of Kafka, once `KafkaProducer.send(...)` is invoked, currently vertx Kafka client provides no way to abort that transaction. We can finish the promise, but it doesn't mean that Kafka producer won't forward that message so that the user can resend it again. The `delivery.timeout.ms` can be set only when we create the _producer_.  In our implementation we reuse the cached producers. Lets say for this method implementation that we create a new Kafka Producer with the given timeout value. But still it is not as simple as changing the value of `delivery.timeout.ms`. Because this value has some prerequisite conditions with other timeout properties namely `request.timeout.ms` and `linger.ms`. I think the user should deal with the Kafka properties directly and we providing an abstraction doesn't make sense to me due to dependeny of one property with that of the others. Since there is already a provision to set the timeout value at the `ApplicationClient` level, IMHO I am not fully convinced that we should also provide on a more granular level in methods.

Reply:
Previously my focus was mainly to ensure that the command is not being sent to the device in case timeout happens. I see  that the interest behind is to protect the client and the client need not wait more than the specified timeout value. It means that there is no guarantee that the command will not be sent to the device, when a request is timed out . I will pull up the method to the `CommandSender` and make use of a timer to implement it.

Reply:
> Since there is already a provision to set the timeout value at the ApplicationClient level, IMHO I am not fully convinced that we should also provide on a more granular level in methods.

Sounds reasonable to me. So FMPOV we should probably simply omit this variant of the method altogether. WDYT?

Reply:
Sounds good to me. I will omit this variant. And also I will update the IT test support to with a _vertx_ timer solution so that this method won't be used there anymore.

Reply:
I have updated and pushed the changes. Would you mind taking a look again?